node/src/components/chain_synchronizer.rs

```diff
@@ -155,347 +147,15 @@ where
     fn handle_fast_sync_result(
         &mut self,
         effect_builder: EffectBuilder<REv>,
-        result: Result<FastSyncOutcome, Error>,
+        result: Result<BlockHeader, Error>,
+        maybe_immediate_switch_block_data: Option<Box<ImmediateSwitchBlockData>>,
     ) -> Effects<Event> {
         self.progress.finish();
         match result {
-            Ok(FastSyncOutcome::ShouldCommitGenesis) => self.commit_genesis(effect_builder),
-            Ok(FastSyncOutcome::ShouldCommitUpgrade {
-                switch_block_header_before_upgrade,
-                is_emergency_upgrade,
-            }) => self.commit_upgrade(
-                effect_builder,
-                switch_block_header_before_upgrade,
-                is_emergency_upgrade,
-            ),
-            Ok(FastSyncOutcome::Synced {
-                highest_block_header,
-            }) => {
+            Ok(highest_block_header) => {
                 self.joining_outcome = Some(JoiningOutcome::Synced {
-                    highest_block_header,
-                });
-                Effects::new()
-            }
-            Err(Error::RetrievedBlockHeaderFromFutureVersion {
-                current_version,
-                block_header_with_future_version,
-            }) => {
-                let future_version = block_header_with_future_version.protocol_version();
-                info!(%current_version, %future_version, "shutting down for upgrade");
-                self.joining_outcome = Some(JoiningOutcome::ShouldExitForUpgrade);
-                Effects::new()
-            }
-            Err(error) => {
-                error!(%error, "failed to sync linear chain");
-                fatal!(effect_builder, "{}", error).ignore()
-            }
-        }
-    }
-
-    fn commit_genesis(&self, effect_builder: EffectBuilder<REv>) -> Effects<Event> {
-        let genesis_timestamp = match self.config.genesis_timestamp() {
-            None => {
-                return fatal!(
-                    effect_builder,
-                    "node started with no trusted hash, no stored blocks, and no genesis timestamp \
-                    in chainspec - specify a trusted hash and restart"
-                )
-                .ignore();
-            }
-            Some(timestamp) => timestamp,
-        };
-
-        let now = Timestamp::now();
-        let era_duration = self.config.era_duration();
-        if now > genesis_timestamp + era_duration {
```

Comment:
Checks like this are absent in the new code. I assume it's by design, just adding a comment to double-check.

Reply:
I think that shouldn't be a problem - if I'm right, this check existed just because before we had immediate switch blocks, there wouldn't have been anything to seed the syncing process in the node if it was started without any blocks, without a trusted hash and after the genesis era. Now we can create the immediate switch block and go forwards from there using finality signatures.

But if I'm missing something and these checks are still necessary, I'll restore them.

Reply:
Per new requirements, we now require nodes started after the genesis timestamp to have a trusted hash - so a check like that has been introduced again in [f47a402](https://github.com/casper-network/casper-node/pull/3280/commits/f47a4023b92e3ecbd2042424ab1ae8e4df51235d)